### PR TITLE
Ri mi start event 2

### PR DESCRIPTION
**Pull request by odowdj**
@padraig-ohiceadha There was a boolean field declared by RI/MI - "isPractice" I'm not 100% sure where to put this, I thought object extensions, but would like your insight please?
Thanks
James

Response from @padraig-ohiceadha for @krisiye 
Is this flag indicating that the test is being taken in free play mode (e.g. as a formative assessment)?

Would we assign tests like this (one way an assigned test could be distinguished from a free play activity is to see if it is associated with an assignment).

### CODE REVIEW

**Comment by krisiye** [comment]
@padraig-ohiceadha @odowdj  - Practice Tests are an attribute of the quiz. Usually the kids will be presented with a few questions that do not affect the scoring so that they get used to the program. They are not going to be assigned outside of the main test. Just part of the same quiz experience.

File: src/test/resources/fixtures/hmh-ri/caliperEventAssessmentCompleted.json [hunk 1]
```diff
@@ -0,0 +1,342 @@
+{
```
Comment (odowdj) [inline src/test/resources/fixtures/hmh-ri/caliperEventAssessmentCompleted.json:1]: Ignore this file please.

**Review by krisiye — COMMENTED** [review]
(not sure what id to put here) - we want that to be the class id right? At the moment there is no way to get that from the launch params. will need enhancements to the LTI tool provider code to make that as part of the launch. We may want to have the class info passed around as part of the LTI launch from Ed and make that available to the launch params in the form of callbacks.

**Comment by kevinmcgarry** [comment]
hi @krisiye - that might be worthwhile doing, not just for this scenario but for numerous other scenarios where content is LTI launched but where we want to capture the Section the action was related to.

File: src/test/resources/fixtures/hmh-ri/caliperEventBasicCreated.json [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        "name": "Too Much Noise",
+        "creators": [
+          {
+            "id": "https://url/to/author/id",
+            "type": "Person",
+            "name": "Ann McGovern"
+          }
+        ],
+        "extensions": {
+          "lexile": 0
```
Comment (odowdj) [inline src/test/resources/fixtures/hmh-ri/caliperEventBasicCreated.json:43]: @padraig-ohiceadha RI/MI have a lexile score associated with the recommended book - do you know if there is a more applicable place inside "Document" for this value otherwise is extensions ok?

**Comment by padraig-ohiceadha** [comment]
There is an emerging best practice which we should consider following on the creation of iris for Caliper where there isn't a de-referenceable URL, at least where we don't already have an established practice (e.g. if we were to apply it to actor:id it should only be if we change the general HMH practice on user identifiers in Caliper messages.
So the 1st option would be to provide a dereferencable URL. The ideal would be if we had a OneRoster server we could use the OneRoster URL for the entity as the Caliper identifier.
Where we don't there is an "IMS identifier namespace" defined where it starts with a prefix of imsglobalrn: and then has a short code to identify the type of identifier, e.g. "or" stands for OneRoster and then another : followed by an indicator of the audience (the set from which these unique ids are drawn), so if we provide a global namespace (not per-tenant) for user ids we might use an actor.id of "imsglobalrn:or:com.hmh:8bd8cecf-ea7c-448e-828f-d60871ba2ab7

**Comment by padraig-ohiceadha** [comment]
Ed uses group a lot, so I think there would be value in passing the group through the launch (as discussed in Slack) and then reporting it in the Caliper events

**Comment by odowdj** [comment]
@padraig-ohiceadha the "group" that you refer to being used in Ed - does it reference the Section that a user belongs to? See https://github.com/hmhco/caliper-common-fixtures/pull/3/files#diff-5acb4eb818ef984723ea3682a17d02e2R46 for my interpretation, thanks

File: src/test/resources/fixtures/hmh-ri/caliperEventAssessmentStarted.json [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    "maxScore":25.0,
+    "version":"1.0"
+  },
+  "generated":{
+    "id":"https://www.hmhco.com/readinginventory/quiz/c35603f7a3434e2cb74ef14d60e70f42/instance/belonging/to/student/userRefId/attempts/1",
+    "type":"Attempt",
+    "assignee":"https://https://www.hmhco.com/ed/users/userRefId", (same as actors id)
+    "assignable":"https://www.hmhco.com/readinginventory/quiz/c35603f7a3434e2cb74ef14d60e70f42/instance/belonging/to/student/userRefId", (same as object id)
+    "count":1,
+    "dateCreated":"2016-11-15T10:15:00.000Z",
```
Comment (padraig-ohiceadha) [inline src/test/resources/fixtures/hmh-ri/caliperEventAssessmentStarted.json]: The dateCreated is the time the attempt was created, which is also just now.

File: src/test/resources/fixtures/hmh-ri/caliperEventAssessmentStarted.json [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      "subOrganizationOf":{
+        "id":"https://url/to/ids/districts/districtRefId",
+        "type":"Organization"
+      }
+    },
+    "roles":[ "Learner" ],
+    "status": "Active",
+    "dateCreated": "2016-08-01T06:00:00.000Z"
+  }
+}
```
Comment (padraig-ohiceadha) [inline src/test/resources/fixtures/hmh-ri/caliperEventAssessmentStarted.json:46]: Should have a session which reflects the test session, and a federatedSession, using a federated session id which comes from the LTI launch (as previously discussed on Slack)

File: src/test/resources/fixtures/hmh-ri/caliperEventAssessmentStarted.json [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  },
+  "eventTime":"2016-11-15T10:15:00.000Z",
+  "edApp":{
+    "id":"https://www.hmhco.com/readinginventory",
+    "type":"SoftwareApplication",
+    "version":"v1"
+  },
+  "group": {
+    "id": "https://url/to/ids/district/school/sections/sectionId1",
+    "type": "CourseSection",
```
Comment (padraig-ohiceadha) [inline src/test/resources/fixtures/hmh-ri/caliperEventAssessmentStarted.json]: If we use the group as a "group", rather than as a class then we'd use a type of "Group" instead of "CourseSection", or could combine them where e.g. the group is a subOrganizationOf the class CourseSection

File: src/test/resources/fixtures/hmh-ri/caliperEventBasicCreated.json (hunk 2 above)
Comment (padraig-ohiceadha) [reply src/test/resources/fixtures/hmh-ri/caliperEventBasicCreated.json:43]: The recommendation is to add lexile levels in Caliper by adding them as learningObjectives.
I know there is a best practices document on this. I'll find it and post it to Slack...

File: src/test/resources/fixtures/hmh-ri/caliperEventAssessmentStarted.json [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    "id":"https://https://www.hmhco.com/ed/users/userRefId",
+    "type":"Person",
+    "name" : "Casandra Rath"
+  },
+  "action":"Started",
+  "object":{
+    "id":"https://www.hmhco.com/readinginventory/quiz/c35603f7a3434e2cb74ef14d60e70f42/instance/belonging/to/student/userRefId",
+    "type":"Assessment",
+    "name":"HMH-RI Quiz Name - Reading Inventory",
+    "dateToStartOn":"2016-11-14T05:00:00.000Z",
```
Comment (padraig-ohiceadha) [inline src/test/resources/fixtures/hmh-ri/caliperEventAssessmentStarted.json]: As an assignable Digital Resource an Assessment has multiple dates available to it.
I'd recommend looking at the full set and seeing if using more of the full set may help address some of our use cases. 
e.g. dateToActivate may indicate when a proctor/invigilator opened the assessment to the candidates, remember than the dateCreated doesn't reflect the data of the assessment content, it's when the assignment was created.

File: src/test/resources/fixtures/hmh-ri/caliperEventAssessmentStarted.json [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  "actor":{
+    "id":"https://https://www.hmhco.com/ed/users/userRefId",
+    "type":"Person",
+    "name" : "Casandra Rath"
+  },
+  "action":"Started",
+  "object":{
+    "id":"https://www.hmhco.com/readinginventory/quiz/c35603f7a3434e2cb74ef14d60e70f42/instance/belonging/to/student/userRefId",
+    "type":"Assessment",
+    "name":"HMH-RI Quiz Name - Reading Inventory",
```
Comment (padraig-ohiceadha) [inline src/test/resources/fixtures/hmh-ri/caliperEventAssessmentStarted.json]: We don't have any items here. Even for an adaptive test, isn't there an initial set of items at least in the test?

**Comment by padraig-ohiceadha** [comment]
I've been thinking that one way to represent the adaptive recommendation of items while the student is taking the test would be to send a basic Event with a object which is the Assessment, with the items array updated to include the recommended items and with an action of Modified (the test the candidate is taking has been modified by the inclusion of additional items on the basis of recommendations returned by the adaptive engine).

File: src/test/resources/fixtures/hmh-ri/caliperEventAssessmentStarted.json (hunk 3 above)
Comment (padraig-ohiceadha) [reply src/test/resources/fixtures/hmh-ri/caliperEventAssessmentStarted.json]: In caliper you get one entry for a group, which is an Organization. Most commonly that group is a CourseSection (which is a class), however it can also be a true group, i.e. a named list of members.
A lot of the new functionality for assessment in Ed for Renaissance Learning is assigning users to groups for assessments, so it seems like a good idea to me to represent the group in the Caliper event.
You don't have to lose the class, if the group is a sub set of the class then the group can be a subOrganization of the CourseSection (group).

File: src/test/resources/fixtures/hmh-ri/caliperEventAssessmentStarted.json (hunk 4 above)
Comment (padraig-ohiceadha) [reply src/test/resources/fixtures/hmh-ri/caliperEventAssessmentStarted.json:46]: The membership object should be the same as for other Caliper events we already handle

File: src/test/resources/fixtures/hmh-ri/caliperEventAssessmentStarted.json (hunk 3 above)
Comment (odowdj) [reply src/test/resources/fixtures/hmh-ri/caliperEventAssessmentStarted.json]: I understand what you are saying - and if there was no group of students within the class, this group would fall back to being just the CourseSection itself right?

File: src/test/resources/fixtures/hmh-ri/caliperEventAssessmentStarted.json (hunk 4 above)
Comment (odowdj) [reply src/test/resources/fixtures/hmh-ri/caliperEventAssessmentStarted.json:46]: Thanks @padraig-ohiceadha if "SessionEvent" comes on board it can be added in to "session". 
Same for "federatedSession", if its a available, I'd imagine it can be added here also

File: src/test/resources/fixtures/hmh-ri/caliperEventAssessmentStarted.json (hunk 3 above)
Comment (padraig-ohiceadha) [reply src/test/resources/fixtures/hmh-ri/caliperEventAssessmentStarted.json]: Exactly

File: src/test/resources/fixtures/hmh-ri/caliperEventBasicCreated.json [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  "@context": "http://purl.imsglobal.org/ctx/caliper/v1p1",
+  "id":"urn:uuid:uniqueIdPerEvent",
+  "type": "Event",
+  "actor": {
+    "id": "https://url/to/lexileService",
+    "type": "SoftwareApplication",
+    "version": "v1"
+  },
+  "action": "Created",
+  "object": {
```
Comment (padraig-ohiceadha) [inline src/test/resources/fixtures/hmh-ri/caliperEventBasicCreated.json:11]: The Created action isn't really being performed on the Attempt, the object it being performed on is the DigitalResourceCollection. So I'd expect what you have in the generated property be used for object and provide no generated property. I think it is fine though to reference the Attempt which led to this resource being created in the referrer property.
The value of mediaType has to be a valid media type, like application/pdf, text/html etc.

Comment (odowdj) [reply src/test/resources/fixtures/hmh-ri/caliperEventBasicCreated.json:11]: Thanks Padraig, a couple of things...

DigitalResourceCollection is the list of recommended books... I'm thinking that this CreatedEvent is a list of these recommendations that were created by the score service, generated from the students specific assessment.
Yes, we can take out the attempt but I think it should be replaced with a reference the specific assessment itself i.e. the values I have here https://github.com/hmhco/caliper-common-fixtures/pull/3/files#diff-804b2bc03ee3fdd1d976f09fc3e1bbfbR20 Thoughts?

**Comment by padraig-ohiceadha** [comment]
Hi James,

     But we already have a reference to the Attempt in the referrer, and the Attempt is linked to the assessmentTest.

         All the best,

               Padraig

**Comment by odowdj** [comment]
Hi @padraig-ohiceadha 
The referrer points back to the score, because there must be an object, the list of recommendations isn't the object, the recommendations have been generated from the object - hence the object is the assessment, the generated data is the recommendations and the referrer is the score

File: src/test/resources/fixtures/hmh-ri/caliperEventGradeGraded.json [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    "comment": "The lexile score from Test Quiz # c35603f7a3434e2cb74ef14d60e70f42 taken on date/time for student userRefId",
+    "dateCreated": "2016-11-15T10:56:00.000Z"
+  },
+  "group": {
+    "id": "https://url/to/ids/district/school/sections/sectionId1",
+    "type": "CourseSection",
+    "courseNumber": "CPS 435-01",
+    "academicSession": "Fall 2016"
+  },
+  "membership":{
```
Comment (odowdj) [inline src/test/resources/fixtures/hmh-ri/caliperEventGradeGraded.json]: @padraig-ohiceadha should I omit the group and membership because the actor is the lexile service?

**Comment by padraig-ohiceadha** [comment]
Hi James,

    The list of recommendations is part of the object. The object is the object being created. So it’s the DigitalResourceCollection which hold the array of items which are the individual documents being recommended.

        All the best,

             Padraig

**Comment by krisiye** [comment]
Most of what we have looks good to me. @odowdj and I discussed briefly last week that we could use the Assessment/Activity refID as the Assessment Object ID (instead of the RI quiz ID - which is mostly used for internal tracking by the RI/lexile service). I'll be doing some additional work to persist the activityID alongside a quiz for correlations.

File: src/test/resources/fixtures/hmh-ri/caliperEventGradeGraded.json [hunk 10]
```diff
@@ -1,60 +1,54 @@
...
     "type": "Score",
-    "attempt": "https://www.hmhco.com/readinginventory/quiz/c35603f7a3434e2cb74ef14d60e70f42/instance/belonging/to/student/userRefId/attempts/completedAttemptId",
-    "maxScore": 15.0,
-    "scoreGiven": -487.0,
-    "scoredBy": "https://url/to/lexileService", (actor id)
-    "comment": "The lexile score from Test Quiz # c35603f7a3434e2cb74ef14d60e70f42 taken on date/time for student userRefId",
+    "attempt": "urn:uuid:c35603f7a3434e2cb74ef14d60e70f42", // ri-quiz-id
+    "maxScore": 1700,
+    "scoreGiven": 800,
+    "scoredBy": "https://www.hmhco.com/ReadingInventory",
```
Comment (odowdj) [inline src/test/resources/fixtures/hmh-ri/caliperEventGradeGraded.json]: this should point to the lexile service right?

File: src/test/resources/fixtures/hmh-ri/caliperEventGradeGraded.json [hunk 11]
```diff
@@ -1,60 +1,54 @@
 {
   "@context": "http://purl.imsglobal.org/ctx/caliper/v1p1",
-  "id":"urn:uuid:uniqueIdPerEvent",
+  "id":"urn:uuid:a50ca17f-5971-47bb-8fca-4e6e6879001d",
   "type": "GradeEvent",
   "actor": {
-    "id": "https://url/to/lexileService",
+    "id": "https://www.hmhco.com/ReadingInventory",
```
Comment (odowdj) [inline src/test/resources/fixtures/hmh-ri/caliperEventGradeGraded.json]: The actor id here should be the lexile service because that's the person/application that generated this event - can you confirm my thinking @padraig-ohiceadha?

Secondly, because Ed Analytics uses Actor id for its partition key, it means until this event has student user id as the actor id, it won't get processed correctly.... so my question for @padraig-ohiceadha what can we do in this scenario? Would it be ok to have actor as the student and 'scoredBy' as the lexile service https://github.com/hmhco/caliper-common-fixtures/pull/3/files/8031736dcab941210bf9e30217ddc4f2e218f4bd..ef6e9e8d2b3a3ffe3d96cb2e5fec1138365d313c#diff-cda1d6654d25effaed17317c5a5e8f44R40 ?

Comment (padraig-ohiceadha) [reply src/test/resources/fixtures/hmh-ri/caliperEventGradeGraded.json]: Yes, the actor should be the Software Application which generated the event.
We do have flexibility in how granular we define applications, we could say everything is ED.
However as ED has more than one scoring service (e.g. Learnosity is also scoring, and we'll hopefully have QTI scoring in the future) and Common Patterns scores also (and is working on Caliper at the moment also), it would be useful to be able to see what particular scoring service was used to score this Attempt

Comment (padraig-ohiceadha) [reply src/test/resources/fixtures/hmh-ri/caliperEventGradeGraded.json]: Unfortunately the student is not an actor here. Unless we have an edge case where they are scoring their own results.
The actor is that who/which performs the action which leads to the generated.
As most analyses of this event will be more interested in who the assignee is than who the actor is I suppose it makes sense to use the assignee as the partition key.
Remember though that when manual scoring comes in to play there will be two people involved and you may be interested in doing some analytics by scorer also.
There is a general pattern of events with 2 human actors (you often also get in communication and collaboration), it might be an idea to discuss as a team how best to handle such events

Comment (krisiye) [reply src/test/resources/fixtures/hmh-ri/caliperEventGradeGraded.json]: Just to clarify terminology here - ReadingInventory is the Lexile Service.

File: src/test/resources/fixtures/hmh-ri/caliperEventGradeGraded.json (hunk 10 above)
Comment (krisiye) [reply src/test/resources/fixtures/hmh-ri/caliperEventGradeGraded.json]: I made the attempt use the quiz ID. Scored By ReadingInventory/Lexile Service

File: src/test/resources/fixtures/hmh-ri/caliperEventGradeGraded.json (hunk 11 above)
Comment (odowdj) [reply src/test/resources/fixtures/hmh-ri/caliperEventGradeGraded.json]: thanks 👍

**Review by odowdj — COMMENTED** [review]
lftm

File: src/test/resources/fixtures/hmh-ri/caliperEventAssessmentStarted.json [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    "id":"urn:uuid:c050e852-5edb-4743-92d1-b53466de3a5f", // assignment activity id
+    "type":"Assessment",
+    "name":"ReadingInventory",
+    "maxAttempts":1,
+    "maxSubmits":1,
+    "maxScore":25.0,
+    "version":"1.0"
+  },
+  "generated":{
+    "id":"urn:uuid:c35603f7a3434e2cb74ef14d60e70f42", // ri-quiz id
```
Comment (odowdj) [inline src/test/resources/fixtures/hmh-ri/caliperEventAssessmentStarted.json:21]: Hi @krisiye, the comment here says ri-quiz id - but I don't think that's correct, this is meant to be an attempt id

**Review by krisiye — APPROVED** [review]
